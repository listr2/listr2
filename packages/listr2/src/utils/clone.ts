import rfdc from 'rfdc'

const clone = rfdc({ circles: true })

/**
 * Deep clones a object in the easiest manner.
 */
// TODO: swap this with structured clone whenever we migrate to node 18
export function cloneObject<T extends Record<PropertyKey, any>>(obj: T): T {
  return clone(obj)
}
