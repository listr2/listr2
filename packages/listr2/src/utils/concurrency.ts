// adapted from https://github.com/fabiospampinato/promise-concurrency-limiter
// with MIT license, since it is again ESM only
// we needed to stop on the first failing case unlike the original
export class Concurrency {
  promise: Promise<void>

  private resolve: () => void
  private reject: (error: Error) => void
  private concurrency: number
  private count: number
  private queue: Set<() => void>

  constructor (options: { concurrency: number }) {
    this.concurrency = options.concurrency
    this.count = 0
    this.queue = new Set()
    this.initPromise()
  }

  public add<T>(fn: () => Promise<T>): Promise<T> {
    if (this.count < this.concurrency) {
      return this.run(fn)
    }

    return new Promise<T>((resolve) => {
      const callback = (): void => resolve(this.run(fn))

      this.queue.add(callback)
    })
  }

  private initPromise (): void {
    const { promise, resolve, reject } = this.withResolvers<void>()

    this.promise = promise
    this.resolve = resolve
    this.reject = reject
  }

  private flush (): void {
    for (const callback of this.queue) {
      if (this.count >= this.concurrency) {
        break
      }

      this.queue.delete(callback)

      callback()
    }
  }

  private run<T>(fn: () => Promise<T>): Promise<T> {
    this.count++

    const promise = fn()

    const cleanup = (): void => {
      this.count--

      this.flush()

      if (this.count === 0) {
        this.resolve()
      }
    }

    promise.then(cleanup, (error) => {
      this.queue.clear()
      this.reject(error)
    })

    return promise
  }

  private withResolvers<T>(): { promise: Promise<T>, resolve: (value: T) => void, reject: (error: Error) => void } {
    let resolve: (value: T) => void
    let reject: (error: Error) => void

    const promise = new Promise<T>((res, rej) => {
      resolve = res
      reject = rej
    })

    return {
      promise,
      resolve: resolve!,
      reject: reject!
    }
  }
}
