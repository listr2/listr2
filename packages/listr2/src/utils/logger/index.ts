export * from './logger.constants'
export type * from './logger.interface'
export * from './logger'
