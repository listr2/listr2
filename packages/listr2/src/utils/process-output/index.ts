export type * from './process-output-buffer.interface'
export * from './process-output-buffer'
export * from './process-output-stream'
export * from './process-output'
export type * from './process-output.interface'
export * from './writable'
