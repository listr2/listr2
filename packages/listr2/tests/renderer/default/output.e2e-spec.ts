import { Listr } from '@root'
import type { MockProcessOutput } from '@tests/utils'
import { expectProcessOutputToMatchSnapshot, mockProcessOutput, unmockProcessOutput } from '@tests/utils'

describe('default renderer: output', () => {
  const output: MockProcessOutput = {} as MockProcessOutput

  process.stdout.isTTY = true

  beforeEach(async () => {
    mockProcessOutput(output)
  })

  afterEach(async () => {
    unmockProcessOutput(output)
    jest.clearAllMocks()
  })

  // oYHBlOYGg8juKRkaqigY617eyLbGMuDd
  it('should have persistent output', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: async (_, task): Promise<void> => {
            task.output = 'I will push an output. [0]'

            task.output = 'I will push an output. [1]'

            task.output = 'I will push an output. [2]'
          },
          rendererOptions: { persistentOutput: true }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'oYHBlOYGg8juKRkaqigY617eyLbGMuDd')
  })

  // LGtvNwRVWZWJQuucaYcVDSrjspKwNnVF
  it('should output to output bar with all output', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: async (_, task): Promise<void> => {
            task.output = 'I will push an output. [0]'

            task.output = 'I will push an output. [1]'

            task.output = 'I will push an output. [2]'
          },
          rendererOptions: { outputBar: Infinity }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'LGtvNwRVWZWJQuucaYcVDSrjspKwNnVF')
  })

  // hXDsrvlxponbYraVVuHSWACjDbyUBXkb
  it('should output to output bar with last 2 output', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: async (_, task): Promise<void> => {
            task.output = 'I will push an output. [0]'

            task.output = 'I will push an output. [1]'

            task.output = 'I will push an output. [2]'
          },
          rendererOptions: { outputBar: 2 }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'hXDsrvlxponbYraVVuHSWACjDbyUBXkb')
  })

  // qmiOeXTUyStaFeRgDYVlGoPJMVDbRRuC
  it('should output to output bar with last output', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: async (_, task): Promise<void> => {
            task.output = 'I will push an output. [0]'

            task.output = 'I will push an output. [1]'

            task.output = 'I will push an output. [2]'
          },
          rendererOptions: { outputBar: true }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'qmiOeXTUyStaFeRgDYVlGoPJMVDbRRuC')
  })

  // it('should clear output from output bar', async () => {
  //   await new Listr(
  //     [
  //       {
  //         title: 'This task will execute.',
  //         task: async (_, task): Promise<void> => {
  //           task.output = 'I will push an output. [0]'

  //           task.output = 'I will push an output. [1]'

  //           task.output = 'I will push an output. [2]'

  //           task.output = ''
  //         },
  //         rendererOptions: { outputBar: Infinity, persistentOutput: true }
  //       }
  //     ],
  //     {
  //       concurrent: false,
  //       rendererOptions: { lazy: true }
  //     }
  //   ).run()

  //   expectProcessOutputToMatchSnapshot(output)
  // })

  // ojurBbWZEPenKVeoFcKksBeMbaYWmbGO
  it('should output to nowhere since outputBar and bottomBar is false', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: async (_, task): Promise<void> => {
            task.output = 'I will push an output. [0]'

            task.output = 'I will push an output. [1]'

            task.output = 'I will push an output. [2]'
          },
          rendererOptions: { outputBar: false, bottomBar: false }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'ojurBbWZEPenKVeoFcKksBeMbaYWmbGO')
  })

  // 767BkeBTfR1lrS2ANYYH7CLWPATxqyat
  it('should output to bottom bar', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: async (_, task): Promise<void> => {
            task.output = 'I will push an output. [0]'

            task.output = 'I will push an output. [1]'

            task.output = 'I will push an output. [2]'
          },
          rendererOptions: { bottomBar: Infinity }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, '767BkeBTfR1lrS2ANYYH7CLWPATxqyat')
  })

  // JMCRBo4OtLm7JB3XbcYoRcCdQRiKfPdP
  it('should output to bottom bar from task with no title', async () => {
    await new Listr(
      [
        {
          task: async (_, task): Promise<void> => {
            task.output = 'I will push an output. [0]'

            task.output = 'I will push an output. [1]'

            task.output = 'I will push an output. [2]'
          },
          rendererOptions: { bottomBar: Infinity }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'JMCRBo4OtLm7JB3XbcYoRcCdQRiKfPdP')
  })

  // iI52S3WPytorU9EZKPar2AiBrLAZTVut
  it('should output to bottom bar from task with no title persistently', async () => {
    await new Listr(
      [
        {
          task: async (_, task): Promise<void> => {
            task.output = 'I will push an output. [0]'

            task.output = 'I will push an output. [1]'

            task.output = 'I will push an output. [2]'
          },
          rendererOptions: { bottomBar: Infinity, persistentOutput: true }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'iI52S3WPytorU9EZKPar2AiBrLAZTVut')
  })

  // HhZEM7noGNW4xpgxv4ZtXsPMroPWqrEA
  it('should output to bottom bar persistently', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: async (_, task): Promise<void> => {
            task.output = 'I will push an output. [0]'

            task.output = 'I will push an output. [1]'

            task.output = 'I will push an output. [2]'
          },
          rendererOptions: { bottomBar: Infinity, persistentOutput: true }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'HhZEM7noGNW4xpgxv4ZtXsPMroPWqrEA')
  })

  // d4wXg4yGYak09qivTqgKFZaQJ3PvDZm5
  it('should limit output to bottom bar persistently', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: async (_, task): Promise<void> => {
            task.output = 'I will push an output. [0]'

            task.output = 'I will push an output. [1]'

            task.output = 'I will push an output. [2]'
          },
          rendererOptions: { bottomBar: 2 }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'd4wXg4yGYak09qivTqgKFZaQJ3PvDZm5')
  })

  // NpGb4ry8b6hlK7VkJ1YcXcVibx0k5Sus
  it('should output to bottom bar 2 times at most and delete the prior tasks output when finished', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: async (_, task): Promise<void> => {
            task.output = 'I will push an output. [0]'

            task.output = 'I will push an output. [1]'

            task.output = 'I will push an output. [2]'
          },
          rendererOptions: { bottomBar: true, persistentOutput: false }
        },

        {
          title: 'This task will execute.',
          task: async (_, task): Promise<void> => {
            task.output = 'I will push an output. [0]'

            task.output = 'I will push an output. [1]'

            task.output = 'I will push an output. [2]'
          },
          rendererOptions: { bottomBar: true, persistentOutput: true }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'NpGb4ry8b6hlK7VkJ1YcXcVibx0k5Sus')
  })

  // j7BqsosH97ffW1SQSdkADSm2HnSZQ9nn
  it('should indent long multiline output with persistent output', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: async (_, task): Promise<void> => {
            const start = 'This is a'
            const mid = 'long '
            const end = 'multi line output.'

            task.output = start + mid.repeat(40) + '\n' + mid.repeat(40) + '\n' + mid.repeat(40) + '\n' + mid.repeat(40) + '\n' + mid.repeat(40) + '\n' + end
          },
          rendererOptions: { persistentOutput: true }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'j7BqsosH97ffW1SQSdkADSm2HnSZQ9nn')
  })

  // MjcoXTjPbNRDsgOIbGzvjt7MEaZcmasv
  it.each([ true, false ])('should have persistent output on task fail with bottom bar %s', async (input) => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: async (_, task): Promise<void> => {
            task.output = 'I will push an output. [0]'

            task.output = 'I will push an output. [1]'

            task.output = 'I will push an output. [2]'

            throw new Error('This task has failed')
          },
          rendererOptions: { bottomBar: input, persistentOutput: true }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true },
        exitOnError: false
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'MjcoXTjPbNRDsgOIbGzvjt7MEaZcmasv')
  })
})
