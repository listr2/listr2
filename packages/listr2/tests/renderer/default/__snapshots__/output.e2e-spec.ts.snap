// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`default renderer: output should have persistent output on task fail with bottom bar false: MjcoXTjPbNRDsgOIbGzvjt7MEaZcmasv-exit 1`] = `[]`;

exports[`default renderer: output should have persistent output on task fail with bottom bar false: MjcoXTjPbNRDsgOIbGzvjt7MEaZcmasv-stderr 1`] = `[]`;

exports[`default renderer: output should have persistent output on task fail with bottom bar false: MjcoXTjPbNRDsgOIbGzvjt7MEaZcmasv-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
  ‚Ä∫ I will push an output. [0]
",
  "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚Ä∫ I will push an output. [1]
",
  "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[G‚úñ This task will execute.
  ‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[G‚úñ This task has failed
  ‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[G",
  "‚úñ This task has failed
  ‚Ä∫ I will push an output. [2]
",
  "[?25h",
]
`;

exports[`default renderer: output should have persistent output on task fail with bottom bar true: MjcoXTjPbNRDsgOIbGzvjt7MEaZcmasv-exit 1`] = `[]`;

exports[`default renderer: output should have persistent output on task fail with bottom bar true: MjcoXTjPbNRDsgOIbGzvjt7MEaZcmasv-stderr 1`] = `[]`;

exports[`default renderer: output should have persistent output on task fail with bottom bar true: MjcoXTjPbNRDsgOIbGzvjt7MEaZcmasv-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.

‚Ä∫ I will push an output. [0]
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.

‚Ä∫ I will push an output. [1]
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.

‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚úñ This task will execute.

‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚úñ This task has failed

‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[1A[2K[G",
  "‚úñ This task has failed

‚Ä∫ I will push an output. [2]
",
  "[?25h",
]
`;

exports[`default renderer: output should have persistent output: oYHBlOYGg8juKRkaqigY617eyLbGMuDd-exit 1`] = `[]`;

exports[`default renderer: output should have persistent output: oYHBlOYGg8juKRkaqigY617eyLbGMuDd-stderr 1`] = `[]`;

exports[`default renderer: output should have persistent output: oYHBlOYGg8juKRkaqigY617eyLbGMuDd-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
  ‚Ä∫ I will push an output. [0]
",
  "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚Ä∫ I will push an output. [1]
",
  "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[G‚úî This task will execute.
  ‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[G",
  "‚úî This task will execute.
  ‚Ä∫ I will push an output. [2]
",
  "[?25h",
]
`;

exports[`default renderer: output should indent long multiline output with persistent output: j7BqsosH97ffW1SQSdkADSm2HnSZQ9nn-exit 1`] = `[]`;

exports[`default renderer: output should indent long multiline output with persistent output: j7BqsosH97ffW1SQSdkADSm2HnSZQ9nn-stderr 1`] = `[]`;

exports[`default renderer: output should indent long multiline output with persistent output: j7BqsosH97ffW1SQSdkADSm2HnSZQ9nn-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
  ‚Ä∫ This is along long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long
    multi line output.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
  ‚Ä∫ This is along long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long
    multi line output.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G",
  "‚úî This task will execute.
  ‚Ä∫ This is along long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long long long long long long
    long long long long long long long long long long
    multi line output.
",
  "[?25h",
]
`;

exports[`default renderer: output should limit output to bottom bar persistently: d4wXg4yGYak09qivTqgKFZaQJ3PvDZm5-exit 1`] = `[]`;

exports[`default renderer: output should limit output to bottom bar persistently: d4wXg4yGYak09qivTqgKFZaQJ3PvDZm5-stderr 1`] = `[]`;

exports[`default renderer: output should limit output to bottom bar persistently: d4wXg4yGYak09qivTqgKFZaQJ3PvDZm5-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.

‚Ä∫ I will push an output. [0]
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.

‚Ä∫ I will push an output. [0]
‚Ä∫ I will push an output. [1]
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.

‚Ä∫ I will push an output. [1]
‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
",
  "[2K[1A[2K[G",
  "‚úî This task will execute.
",
  "[?25h",
]
`;

exports[`default renderer: output should output to bottom bar 2 times at most and delete the prior tasks output when finished: NpGb4ry8b6hlK7VkJ1YcXcVibx0k5Sus-exit 1`] = `[]`;

exports[`default renderer: output should output to bottom bar 2 times at most and delete the prior tasks output when finished: NpGb4ry8b6hlK7VkJ1YcXcVibx0k5Sus-stderr 1`] = `[]`;

exports[`default renderer: output should output to bottom bar 2 times at most and delete the prior tasks output when finished: NpGb4ry8b6hlK7VkJ1YcXcVibx0k5Sus-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute.
",
  "[2K[1A[2K[G‚óº This task will execute.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
‚óº This task will execute.

‚Ä∫ I will push an output. [0]
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
‚óº This task will execute.

‚Ä∫ I will push an output. [1]
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
‚óº This task will execute.

‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚ùØ This task will execute.
",
  "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚ùØ This task will execute.

‚Ä∫ I will push an output. [0]
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚ùØ This task will execute.

‚Ä∫ I will push an output. [1]
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚ùØ This task will execute.

‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚úî This task will execute.

‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G",
  "‚úî This task will execute.
‚úî This task will execute.

‚Ä∫ I will push an output. [2]
",
  "[?25h",
]
`;

exports[`default renderer: output should output to bottom bar from task with no title persistently: iI52S3WPytorU9EZKPar2AiBrLAZTVut-exit 1`] = `[]`;

exports[`default renderer: output should output to bottom bar from task with no title persistently: iI52S3WPytorU9EZKPar2AiBrLAZTVut-stderr 1`] = `[]`;

exports[`default renderer: output should output to bottom bar from task with no title persistently: iI52S3WPytorU9EZKPar2AiBrLAZTVut-stdout 1`] = `
[
  "[?25l",
  "
",
  "[2K[1A[2K[G‚Ä∫ I will push an output. [0]
",
  "[2K[1A[2K[G‚Ä∫ I will push an output. [0]
‚Ä∫ I will push an output. [1]
",
  "[2K[1A[2K[1A[2K[G‚Ä∫ I will push an output. [0]
‚Ä∫ I will push an output. [1]
‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[1A[2K[G",
  "‚Ä∫ I will push an output. [0]
‚Ä∫ I will push an output. [1]
‚Ä∫ I will push an output. [2]
",
  "[?25h",
]
`;

exports[`default renderer: output should output to bottom bar from task with no title: JMCRBo4OtLm7JB3XbcYoRcCdQRiKfPdP-exit 1`] = `[]`;

exports[`default renderer: output should output to bottom bar from task with no title: JMCRBo4OtLm7JB3XbcYoRcCdQRiKfPdP-stderr 1`] = `[]`;

exports[`default renderer: output should output to bottom bar from task with no title: JMCRBo4OtLm7JB3XbcYoRcCdQRiKfPdP-stdout 1`] = `
[
  "[?25l",
  "
",
  "[2K[1A[2K[G‚Ä∫ I will push an output. [0]
",
  "[2K[1A[2K[G‚Ä∫ I will push an output. [0]
‚Ä∫ I will push an output. [1]
",
  "[2K[1A[2K[1A[2K[G‚Ä∫ I will push an output. [0]
‚Ä∫ I will push an output. [1]
‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[1A[2K[G
",
  "[2K[1A[2K[G",
  "
",
  "[?25h",
]
`;

exports[`default renderer: output should output to bottom bar persistently: HhZEM7noGNW4xpgxv4ZtXsPMroPWqrEA-exit 1`] = `[]`;

exports[`default renderer: output should output to bottom bar persistently: HhZEM7noGNW4xpgxv4ZtXsPMroPWqrEA-stderr 1`] = `[]`;

exports[`default renderer: output should output to bottom bar persistently: HhZEM7noGNW4xpgxv4ZtXsPMroPWqrEA-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.

‚Ä∫ I will push an output. [0]
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.

‚Ä∫ I will push an output. [0]
‚Ä∫ I will push an output. [1]
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.

‚Ä∫ I will push an output. [0]
‚Ä∫ I will push an output. [1]
‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.

‚Ä∫ I will push an output. [0]
‚Ä∫ I will push an output. [1]
‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G",
  "‚úî This task will execute.

‚Ä∫ I will push an output. [0]
‚Ä∫ I will push an output. [1]
‚Ä∫ I will push an output. [2]
",
  "[?25h",
]
`;

exports[`default renderer: output should output to bottom bar: 767BkeBTfR1lrS2ANYYH7CLWPATxqyat-exit 1`] = `[]`;

exports[`default renderer: output should output to bottom bar: 767BkeBTfR1lrS2ANYYH7CLWPATxqyat-stderr 1`] = `[]`;

exports[`default renderer: output should output to bottom bar: 767BkeBTfR1lrS2ANYYH7CLWPATxqyat-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.

‚Ä∫ I will push an output. [0]
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.

‚Ä∫ I will push an output. [0]
‚Ä∫ I will push an output. [1]
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.

‚Ä∫ I will push an output. [0]
‚Ä∫ I will push an output. [1]
‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
",
  "[2K[1A[2K[G",
  "‚úî This task will execute.
",
  "[?25h",
]
`;

exports[`default renderer: output should output to nowhere since outputBar and bottomBar is false: ojurBbWZEPenKVeoFcKksBeMbaYWmbGO-exit 1`] = `[]`;

exports[`default renderer: output should output to nowhere since outputBar and bottomBar is false: ojurBbWZEPenKVeoFcKksBeMbaYWmbGO-stderr 1`] = `[]`;

exports[`default renderer: output should output to nowhere since outputBar and bottomBar is false: ojurBbWZEPenKVeoFcKksBeMbaYWmbGO-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
",
  "[2K[1A[2K[G‚úî This task will execute.
",
  "[2K[1A[2K[G",
  "‚úî This task will execute.
",
  "[?25h",
]
`;

exports[`default renderer: output should output to output bar with all output: LGtvNwRVWZWJQuucaYcVDSrjspKwNnVF-exit 1`] = `[]`;

exports[`default renderer: output should output to output bar with all output: LGtvNwRVWZWJQuucaYcVDSrjspKwNnVF-stderr 1`] = `[]`;

exports[`default renderer: output should output to output bar with all output: LGtvNwRVWZWJQuucaYcVDSrjspKwNnVF-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
  ‚Ä∫ I will push an output. [0]
",
  "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚Ä∫ I will push an output. [0]
  ‚Ä∫ I will push an output. [1]
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚Ä∫ I will push an output. [0]
  ‚Ä∫ I will push an output. [1]
  ‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
",
  "[2K[1A[2K[G",
  "‚úî This task will execute.
",
  "[?25h",
]
`;

exports[`default renderer: output should output to output bar with last 2 output: hXDsrvlxponbYraVVuHSWACjDbyUBXkb-exit 1`] = `[]`;

exports[`default renderer: output should output to output bar with last 2 output: hXDsrvlxponbYraVVuHSWACjDbyUBXkb-stderr 1`] = `[]`;

exports[`default renderer: output should output to output bar with last 2 output: hXDsrvlxponbYraVVuHSWACjDbyUBXkb-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
  ‚Ä∫ I will push an output. [0]
",
  "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚Ä∫ I will push an output. [0]
  ‚Ä∫ I will push an output. [1]
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚Ä∫ I will push an output. [1]
  ‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
",
  "[2K[1A[2K[G",
  "‚úî This task will execute.
",
  "[?25h",
]
`;

exports[`default renderer: output should output to output bar with last output: qmiOeXTUyStaFeRgDYVlGoPJMVDbRRuC-exit 1`] = `[]`;

exports[`default renderer: output should output to output bar with last output: qmiOeXTUyStaFeRgDYVlGoPJMVDbRRuC-stderr 1`] = `[]`;

exports[`default renderer: output should output to output bar with last output: qmiOeXTUyStaFeRgDYVlGoPJMVDbRRuC-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
  ‚Ä∫ I will push an output. [0]
",
  "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚Ä∫ I will push an output. [1]
",
  "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚Ä∫ I will push an output. [2]
",
  "[2K[1A[2K[1A[2K[G‚úî This task will execute.
",
  "[2K[1A[2K[G",
  "‚úî This task will execute.
",
  "[?25h",
]
`;
